elf_reader.go

```diff
@@ -377,6 +405,61 @@ outer:
 	return nil
 }
 
+func (ec *elfCode) sectionName(idx int) (string, error) {
+	if idx >= len(ec.Sections) {
+		return "", fmt.Errorf("out of bounds section index")
+	}
+	return ec.Sections[idx].Name, nil
+}
+
+func (ec *elfCode) coreRelocateInstruction(ins *asm.Instruction, res *btf.CORERelocation) error {
+	if res.Poison {
+		ins.Poison()
+		return nil
+	}
+
+	origVal := res.OrigVal
+	newVal := res.NewVal
+
+	switch ins.OpCode.Class() {
+	case asm.ALUClass:
+		fallthrough
+	case asm.ALU64Class:
+		if ins.OpCode.Source() != asm.ImmSource {
+			return fmt.Errorf("invalid relocation instruction %v", ins)
+		}
+		if res.Validate && ins.Constant != int64(origVal) {
+			return fmt.Errorf("immediate value %#x doesn't match %#x", ins.Constant, origVal)
+		}
+		ins.Constant = int64(newVal)
+	case asm.LdXClass:
+		fallthrough
+	case asm.StClass:
+		fallthrough
+	case asm.StXClass:
```

Comment:
Nit: I think Go allows you to do `case asm.LdXClass, asm.StClass, asm.StXClass:`